proposals/module-types/Explainer.md

```diff
@@ -0,0 +1,710 @@
+of module types and instance types.
+
+
+### Summary
+
+To summarize the proposed changes:
+* The `module` field of [`import`] become optional (allowing single-level
+  imports). (How to encode this in the [import section] is an interesting
+  question.)
+* New `module` and `instance` type constructors are added that may be used to
+  define types in the [type section].
+* A new `alias` case is added to type definitions allowing the import of parent
+  modules' type definitions.
+* New `module` and `instance` cases are added to [`importdesc`], with the
+  type indicated by `typeidx`, like `func`.
+* A new `module` section is added which contains a sequence of either module
+  definitions or `alias` definitions (of parents' modules). In terms of the
+  binary format, a nested module definition has the same binary format as a
+  non-nested `.wasm` module. Thus, the [`module`] binary format production
+  becomes recursive.
+* A new `instance` section is added which contains a sequence of instance
+  definitions. Like [`global`] definitions, an `instance` definition contains
+  an `instance` type and an `init` expression which is validated to match.
+* A new `instance.instantiate` instruction is added which is only allowed in
+  `instance` `init` expressions.
+* New `ref.instance` and `ref.module` instructions are added to the general
+  instruction set.
+* New module and instance reference types `(ref (module ...))`, `moduleref`,
+  `(ref (instance ..))` and `instanceref` are added to [`reftype`].
+* New `module` and `instance` cases are added to [`exportdesc`].
+
+
+## Use Cases Revisited
+
+Worked examples of the [above use cases](#use-cases) are given in separate docs:
+* [Link-Time Virtualization](Example-LinkTimeVirtualization.md)
+* [Shared-Everything Dynamic Linking](Example-SharedEverythingDynamicLinking.md)
+
+## Additional Requirements Revisited
+
+Reconsidering the requirements stated [above](#additional-requirements):
+* **No GC dependency**: The lifetimes of nested instances are necessarily the
+  same as the lifetime of their parent instance. Thus, this proposal introduces
+  no new interesting lifetimes compared to the status quo today. Of course, if
+  GC is currently needed on a host to collect dead instances (as is the case,
+  e.g., on the Web), it will continue to. But if a host currently has
+  more-explicitly defined instance lifetimes today, that same scheme will
+  continue to work with this proposal.
+* **Enable AOT compilation**: Nested instance definitions are declarative
+  enough to allow an AOT compiler to trivially connect exports to imports
+  to the same degree as a static linker, assuming it knows the resolution
+  of all module imports. In fact, when all module imports are known, a simple,
+  language-agnostic tool can transpile a `.wasm` module using the features
+  of this proposal into a `.wasm` module using only [multi-memory].
+
+
+## FAQ
+
+### How does this relate to ESM-integration?
+
```

Comment:
Would it make sense to define the semantics of the start section when linking?

In the source, what if a library uses top-level code to initialize some state and a module imports that library. I would expect the initlization code to still be run because some experts might rely on it . Which would be run by the start function.